Replace Serializable with "constructor objects".

krzentner opened the pull request:
A rendered version of the proposal is [here](https://github.com/rlworkgroup/proposals/blob/constructor-objects/constructor_objects.md).
Essentially, the goal here is to replace the magic of `Serializable`, with a much more explicit, but not less verbose, mechanism for producing instances of non-picklable types.

I would like explicitly like feedback on the naming of things, the general approach, the API, and alternative proposals.

krzentner committed:
Replace Serializable with "constructor objects".

krzentner committed:
Slightly improve constructor_objects.md formatting

vitchyr commented:
Can you explain the issue you have right now with `Serializable`?

One reason that I like `Serializable` is that it's relatively modular. It's not like everything passed into a `Serializable` class needs to be `Serializable` itself; they just need to be pickle-able (but I think that's a reasonable assumption anyways). You're free to use it when you want to, and not use it when you don't. For example, in your `GatherEnv` example, is there a reason you can't just serialize the environment directly (without making it subclass `Serializable` and without the "constructor objects" logic)? Python already provides a default `__get/setstate__` implementation that basically just saves `__dict__`. 

`Serializable` is also relatively straight-forward to extend. `Parameterized` and `NormalizedEnv` are great examples of that. It forces the code writer to list what parameters they want to save (quite literally, it forces the code writer to implement `__getstate__`).

# Re: proposal
I think that while the proposed alternative is clever, I'm not sure it optimizes for simplicity and easy-of-use. This partial/lazy evaluation proposal reminds me of the stub-logic in `rllab`, and most people I've talked to agree that it was a nightmare.

> Constructor objects are just pickleable objects that can be invoked to construct instances of their corresponding type.

This logic is effectively implemented inside `Serializable`:
```
class Serializable:
    # ...
    def __setstate__(self, d):
        out = type(self)(*d["__args"], **d["__kwargs"])
        self.__dict__.update(out.__dict__)
```
What's the motivation for capturing this logic inside a separate object? I'm not sure I see why this is more simple and/more explicit.

# Warning/side note
I take the practice that `Serializable.quick_init` should be called as the **first line** in the function. This ensures that `Serializable` captures the variables passed into `__init__`, without modification from the class. For example, if you do:
```
class Foo(Serializable):
  def __init__(self, lr):
    lr = lr * 10  # Not sure why you'd do this...
    self.quick_init(locals())
    self.lr = lr
```
then this will save `lr * 10` into the dictionary. Then when you load the policy, `lr` will be multiplied by `100` instead of `10`!

This also makes it so that what is saved is independent of the implementation inside of `__init__`. In other words,

```
class Foo(Serializable):
  def __init__(self, lr):
    # modify lr
    self.quick_init(locals()) # <-- state saved depends on code
```
versus
```
class Foo(Serializable):
  def __init__(self, lr):
    self.quick_init(locals()) # <-- state saved is implementation-agnostic
    # modify lr
```

Is there a reason you call `quick_init` last? That could be the reason why `Serializable` seems like such a headache.

krzentner commented:
Hi Vitchyr,

Thanks for the quick feedback. If you don't mind, I'm going to answer your questions in reverse order.

### Calling `quick_init` last.
I'm not completely sure what reason for calling `quick_init` last is, but I suspect it's because that's the easiest way to make objects inheriting `Serializable` indirectly pickle correctly. Almost every environment in garage calls `quick_init` last, so it's clearly a common pattern for us.
It appears to have been introduced in [this commit](https://github.com/rlworkgroup/garage/commit/58562192d12d7ee1068081e14509d74f96cfcdcc), but it's possible there are earlier references. @gonzaiva, why did you put `quick_init` last?

It's worth mentioning that even if `quick_init` is called first, if the parameters to the constructor are mutated then the same type of bug can occur.

### Issues we're having with `Serializable`
I think this case, itself, explains one major issue we have with `Serializable`. Developers coming to garage do not intuitively understand what `Serializable` is doing, and in many cases put it on normal picklable objects ([1](https://github.com/rlworkgroup/garage/blob/master/garage/envs/env_spec.py#L4), [2](https://github.com/rlworkgroup/garage/blob/master/garage/misc/instrument.py#L27), [3](https://github.com/rlworkgroup/garage/blob/master/garage/exploration_strategies/ou_strategy.py#L12), ...). Since they don't understand what it's doing, they end up calling `quick_init` at the wrong time, expecting their objects to be stateful despite using `Serializable`, or totally [working around](https://github.com/rlworkgroup/garage/blob/master/garage/theano/sampler/replay_pool.py#L70) `Serializable`. I certainly think that `Serializable` is a pretty neat trick, but it's usually being used incorrectly (at least in garage).

Your point about `Serializable` being relatively modular is well taken. A definite weakness of the current proposal is that `constructed_by` makes constructing its target normally very difficult. Perhaps constructor objects should call a different method (`construct`?), so that `__init__` can still be implemented to take arguments. In comparison, `Serializable` only affects the pickling behaviour, which has much more restricted effects on the object's use.

### Stub Logic
On the topic of stub logic, I think this design is totally different from stub. What I've described here is basically just a very lightweight factory interface, which is quite different from creating object wrappers that intercept all field and method accesses. In particular, this proposal doesn't wrap anything at all, nor does it create any type of field interception.

### Implicit vs. Explicit
I should make more clear why I think this proposal is more explicit than `Serializable`. Inheriting from `Serializable` in a class `A` has the following effects:
 - Allows instances of `A` to be pickled if their **constructor arguments** can be pickled, instead of if their **fields** can be pickled.
 - Causes mutations to instances of `A` to not be pickled, unless those mutations actually indirectly affect the constructor arguments. In other words, it implies that `A` is a  "stateless type."
 - Causes all classes inheriting from `A` (directly or indirectly) to become "stateless types."
 - If `A` will be inherited from, then either `locals()` need to be passed up the `__init__` chain, or `quick_init()` needs to be called last (which has its own issues, as you've described).
 - Argument's to `A`'s `__init__` should not be modified before calling `quick_init()`, and should never be mutated.

From looking at how people are using `Serializable`, it seems to me that none of the above points are clear to them. The interface they see to `Serializable` is just inheriting from it and calling `quick_init()`, and neither of those things obviously imply any of the above properties. Consequently, they do not expect `A` to be stateless, and do not understand the consequences of exactly how `quick_init()` is called.

In comparison, constructor objects make these properties very clear:
 - The constructor object is picklable, just like any other picklable object. The constructed object need not be picklable.
 - Mutations to the constructor object affect the constructor object, and mutations to the constructed object affect the constructed object. In other words, there's nothing surprising there. Nothing is forced to be stateless.
 - Inheriting from a constructed object basically requires writing a new constructor object, which is enforced by the decorator. The ease of use here could be improved, but the behaviour is unsurprising and there aren't any easy mistakes to make.

The fact that all of these properties are clear from the interface is why I say constructor objects are "more explicit", although they're not really any more verbose.

### Modularity
Constructor objects, as proposed, clearly have two major weaknesses relative to `Serializable`, both of which have to do with modularity in some sense.

Firstly, constructor objects as proposed make constructing objects for use locally fairly awkward. One needs to do something like `GatherEnvCon(3, 2)()`. However, fixing this problem is quite easy. Simply have `constructed_by` call a different method from `__init__` (e.g. `construct`), and have it generate an `__init__` which does the same thing as e.g. `GatherEnvCon(3, 2)()`.

The bigger issue, which is fundamental to the design, is that constructor objects are making "statelessness" explicit at the interface level.
Currently, if an environment is passed to a function, then that function can choose to either use that environment directly (i.e. modify its state), or use it as a template for creating new environments (i.e. using it "statelessly").
With this proposal, what objects will be used as templates (instead of used directly) becomes part of the interface.

A practical consequence of this is that functions which merely receive instances (which are supposed to be used directly), cannot then be modified to create new instances. From my perspective, this makes  state management clearer, since there's no chance an object was used directly (i.e. had its state modified) before it was used "statelessly." Since that type of interaction seems likely to be a bug, I see this restriction as a good thing.

The other consequence of this is that using those ("stateless") parameters may be somewhat more awkward, if the instances don't really need a constructor. For example, under this proposal we would probably convert the environments to use constructor objects. Consequently, passing a picklable, already constructed object to an interface expecting a constructor object will be more difficult. Most likely, we would expose a `Preconstructed` function (name subject to change), which returns a constructor that just copies the input.

I don't think this is actually a major problem in practice, since most types are usually either almost always picklable or almost never. For example, environments are almost never naively picklable.

On the topic of extending `Serializable`, I definitely think that the pickle interfaces `__getstate__` and `__setstate__` are very powerful. However, it's unclear to me what the value of using `Serializable` is if those methods are redefined anyways.

Hopefully this answers your questions in enough detail. Sorry for the wall of text.
I'm not trying to replace `Serializable` because it's inherently a bad idea, I'm trying to replace it because it usually has been used wrong.

krzentner committed:
Respond to suggestions from vitchyr.

Replace `__init__` in constructed objects with `construct`. This allows
constructed objects to also be created normally, by defining `__init__`.
This also fixes pylint, since it no longer expects an `__init__` super
call. Finally, this fixes `attr`, since it defines its own `__init__`.

Since `__init__` is automatically defined, allow `construct` to
similarly not be defined.

Add a `Preconstructed` constructor object, for treating picklable
objects as constructor objects by cloning them.

hartikainen commented on constructor_objects.md:
````diff
@@ -0,0 +1,105 @@
+In experiment file (at least in principle):
+
+```python
+run_experiment(..., env_con=GatherEnvCon(3, 2))
````
Should this be `GatherEnv(3, 2)`?

krzentner replied:
Nope, it should definitely be `GatherEnvCon(3, 2)`.

hartikainen replied:
Yep, my bad. I commented a bit hastily before actually reading the whole implementation.

hartikainen commented:
I'd also like to add my two cents. I think most of the things pointed out comments above are all valid. Overall, I feel like Serializable should be replaced with more explicit serialization, but at the same time I'm not fully convinced (though I might be wrong here) that the Constructor pattern is the best way to achieve this.

I do especially agree  with the "Implicit vs. Explicit" argument made by @krzentner above. Some of the biggest painpoints of Serializable, for me, have been the side-effects of its implicity, especially when subclassing things. Nowadays, I'm pretty comfortable using it because I know exactly how it works and how it's meant to be used, but I think it took quite many bugs to get there.

My worry about the proposed implementation is that it undermines the biggest benefit of Serializable, i.e. that the user doesn't have to implement custom methods (`__setstate__` and `__getstate__`) in order to make stateless objects pickleable. With the new proposal, the user would have to implement a separate constructor class, which sounds like it has potential to become nearly as cumbersome as implementing the `__setstate__` and `__getstate__` explicitly. I'm not 100% that this is true, but I think that's something to keep in mind and test before actually committing to it. If I had to choose between writing a custom `__setstate__` and `__getstate__` or equally complicated custom contructor, I think I would actually prefer the former, mainly because it is a more standard way of implementing custom pickle functionality.

Btw, I [reimplemented the serializable class](https://github.com/hartikainen/serializable) a while ago due to couple of bugs in the rllab's Serializable (e.g. handling of kwargs and ported it to python3). Feel free to use the code in whatever way you'd like, in case you decide to stick with the current Serializable and need to rewrite some of the things.

ryanjulian commented:
KR spent a lot of time talking about environments, but most of his comments apply equally to serializing neural network modules.

Other than the reasons KR talked about, the main reason I'm interested in replacing Serializable is that it forces a central "star-shaped" dependency onto any object which needs to interact with the sampler or the checkpoint logger (i.e. most environments and neural network modules). It doesn't even just force the dependency at the top of the class hierarchy, either: you have to call `Serializable.quick_init()` from every single class you want to serialize.

This means that if we split garage without removing `Serializable`, it would become part of virtually every interface between packages.

### KR's interface is actually just `Callable`
The decorator is nice, but the actual interface KR is proposing is just `Callable`. It says that functions which need Serializable representations of complicated objects should instead take a `Callable` which will reify that object.

Today I have to do this to serialize a gym environment
```python
class GymEnv(Serializable, gym.Wrapper):

    def __init__(self, name):
        Serializable.quick_init(self, locals())
        self.env = gym.make(name)

run_experiment(..., env=GymEnv('Cartpole-v1'))
```

For instance, using KR's example, I could run an experiment on a gym environment without ever having to write a `Serializable` wrapper.

```python
run_experiment(..., env_con=functools.partial(gym.make, 'Cartpole-v1'))
```

### Motivating Example: Splitting algorithms from the experiment runner
Imagine we want to move the parallel sampler, checkpoint logging, live plotting, and other systems stuff into a separate package (call it `runner`), away from the algorithms and neural networks (call it `algos`). The premise of the `runner` package is that anyone should be able to use those great experiment tools with their own algorithms by installing the `runner` package, rather than being forced to work in a fork of `garage`.

If we did that today, `runner` could only sample from environments and policies which implement `Serializable`. Both `runner` and `algos` would need `Serializable` in their import path.

If we decided to keep `Serializable` there are several choices for the dependency graph, none of which are very appealing to me:
* Copy `Serializable` into every new package and pray
    I don't even want to contemplate this one.

* Keep `Serializable` in the `algo` package
    [algo] <-- [runner]
    This is bad. Now `runner` depends on `algo`, but we want there to be many `algo` packages (e.g. tf_baselines, torch_baselines, my_cool_new_research_repo). So now we have to pick a "golden" `algo` package from where everything else imports `Serializable`.

* Keep `Serializable` in the `runner` package 
    [runner] <-- [algo]
    Now all algorithms and models in `algo` have to be designed with a tightly-coupled dependency to `runner`. Because `quick_init()` must appear in the constructor of every object to be serialized, the `Serializable` class from `runner` will appear in a huge percentage of the `algo` classes (policies, q_functions, etc.).

* Make a separate package for `Serializable` and have everything depend on that 
    [algos_tf] --> [serializable] <-- [runner], [algos_torch] --> [serializable], etc.
    This is better than above, but is a lot of work. It also doesn't solve the "I don't want to rewrite tons of my algorithm to use your runner" problem, it merely renames the package where the interface lives.

### It only gets a little better if you narrow the interface
Some interface between `runner` and `algo` clearly has to be defined. Currently in garage that interface is quite large, because the sampler/runner/checkpointer directly pickles/unpickles policies, Q functions, baselines, etc.

We should expect the situation would be made better by forcing a narrow interface between `runner` and `algo` and chasing `Serializable` out of the algorithm implementation (e.g. policies objects no longer implement Serializable, just whole algorithms).

lagom and rllib call this the `Agent` interface and I will write a separate proposal for one soon.

```python
class Agent(Serializable):

    def train_once(self, data):
        raise NotImplementedError

    def get_actions(self, inputs):
        raise NotImplementedError

    def get_diagnostics(self):
        raise NotImplementedError

    # implements Serializable
    def __get_state__(self):
        raise NotImplementedError

    # implements Serializable
    def __set_state__(self, data):
        raise NotImplementedError

    # implements Parameterized
    def get_params(self):
        raise NotImplementedError

    # implements Parameterized
    def set_params(self, param_values):
        raise NotImplemenetedError
```

Now `runner` can clone an agent (e.g. for a sampler worker process) by pickling/unpickling according to `Serializable` and update that agent's state by transmitting only parameters (`get_params()` and `set_params()`).

However, this doesn't really help much. To implement this `__get_state__()`, the author of `Agent` still needs to somehow collect and return all of the constructor args for all of its internal objects (e.g. policies, Q functions, baselines). And similarly for `__set_state__()`, the author needs to reify all of those internal objects from just the arguments data.

```python
class PPOAgent(Agent):

    def __init__(self, policy, baseline, ...):
        Serializable.quick_init(locals())
        
        self._policy = policy
        self._baseline = baseline

    def __get_state__(self):
        # policy and baseline have already been constructed, so I have to ask them how to recreate themselves
        return {
            'policy': self._policy.i_dont_know(),
            'baseline': self._baseline.i_dont_know(),
        }
```

If they use `Serializable` to provide `i_dont_know()`, then `Serializable` ends up deep into the entire class hierarchy of `Agent`, and the interface simplified nothing about this problem. This is the current state of the world.

### It gets a lot better if you separate object specifications from their instances (KR's idea)
If they use KR's idea (or something like it) then the problem could be much simpler.

```python
class PPOAgentSpec:
    
    def __init__(self, policy_spec, baseline_spec):
        self._policy_spec = policy_spec
        self._baseline_spec = baseline_spec


@constructed_by(PPOAgentSpec)
class PPOAgent(Agent):

        def construct(self):
            self._policy = self._policy_spec()
            self._baseline = self._baseline_spec()
            # let's assume PPOAgentSpec stores itself at self._spec

        def __get_state__(self):
            # everything started as a stateless spec anyway, just keep that and return it.
            return self._spec
```

### If you have specs, you don't really need `Serializable` at all
If you're willing to buy my example above, then there's really no need for `Serializable` to be a part of the `Agent` interface either.

```python
class Agent:

    def train_once(self, data):
        raise NotImplementedError

    def get_actions(self, inputs):
        raise NotImplementedError

    def get_diagnostics(self):
        raise NotImplementedError

    # replaces __get_state__()
    # no __set_state__() needed
    @property
    def spec(self):
        raise NotImplementedError

    # implements Parameterized
    def get_params(self):
        raise NotImplementedError

    # implements Parameterized
    def set_params(self, param_values):
        raise NotImplemenetedError

    # now pickling can even have a straightforward implementation
    def __get_state__(self):
        return {'spec': self.spec, 'params': self.get_params()}

    def __set_state__(self, data):
        out = data['spec']()
        self.__dict__.update(out.__dict__)
        self.set_params(data['params'])


class PPOAgentSpec:
    
    def __init__(self, policy_spec, baseline_spec):
        self._policy_spec = policy_spec
        self._baseline_spec = baseline_spec


@constructed_by(PPOAgentSpec)
class PPOAgent(Agent):

        def construct(self):
            self._policy = self._policy_spec()
            self._baseline = self._baseline_spec()
            # let's assume PPOAgentSpec stores itself at self._spec

        @property
        def spec(self):
            return self._spec
```

In this example, the only types in the interface between `runner` and `algo` are `Agent` and primitives. The `Agent` interface doesn't force any state representation on the `algo` author, other than `spec` must return a pickleable `Callable` which will recreate the Agent. This makes it easy to write an `Agent` wrapper for your algorithm which is best suited to your implementation decisions.

vitchyr commented:
You touch on this at the end, but one question that's running at the back of my head is: **why did you want `Serializable` in the first place?** I imagine it was originally implemented because it can generally be more memory-efficient to save the constructor inputs rather than the contents of `__dict__`, but is it actually worth this headache? (Personally, I also like `Serializable` because I can then later change the code and still be able to de-serialize the object, so long as the constructor interface is the same.)

I don't think any of the code should actually depend `Serializable`. In the `runnner`/`algo` example, you say

> If we did that today, runner could only sample from environments and policies which implement `Serializable`. 

Can you explain this? For me, `Serializable` is not an interface. It's an implementation of the `pickle` interface: it gives a light-weight implementation of `__getstate__` and `__setstate__` (so that you can call `pickle.dump` and `pickle.load`). I'm not sure why `runner` would only be compatible with `Serializable` environments/policies. **Isn't `__getstate__` and `__setstate__` the only interface you need?**


Regardless, I see the point you're making. `Serializable` effectively saves the kwargs for a factory, and so any modified state is by default forgotten. At this point, I "just know" this and avoid these bugs, but I am also acutely aware of how easy it is to forget/annoying it is to figure out. I also see the merit of making this more explicit: save factories, rather than making classes implement factory-based `__setstate__`. I think I agree with Kristian that I'm not sure if the Constructor/factory method is the best alternative.

My reference to the stub logic is that the partial logic might get messy once you have multiple levels of dependencies. With Serializable, it's relatively straightforward to do
```
encoder = Encoder(**encoder_params)
decoder = Decoder(**decoder_params)
vae = VAE(encoder, decoder)
run_experiment(..., vae=vae)
```
but with the factory method, you'll usually need something like
```
run_experiment(..., 
  vae=functools.partial(
    VAE,
    functools.partial(Encoder, encoder_params),
    functools.partial(Deocder, decoder_params),
))
```
Obviously there might be a cleaner way to do this, but my point is that this can get messy rather quickly.

Also, it seems like you'll need a factory for each thing you want to save...but I guess you could default to just saving the entire object by calling `__getstate__` if it doesn't have an associated factory? (But then why not just make `__getstate__`/`__setstate__` **the** interface?)

zuoxingdong commented:
I agree with @vitchyr . This proposal looks great, for most common environment objects they are already picklable, so it would be convenient to simply allow them to be defined without subclassing anything. And `constructed_by` decorator is used for those that are not picklable directly. 

For the neural network related objects, I like the way how PyTorch deals with it in `nn.Module`, and common building blocks like `MLP`, `Policy`, `Agent` could be conveniently subclass a base class with such style. It could often make code much cleaner. 

For example
```python 
feature_networks = FeatureNetworks(...)
action_head = DiagonalGaussianHead(...)
policy = Policy(feature_networks, action_head, ...)
agent = PPOAgent(policy, ....)

agent.num_params
optimizer = Adam(agent.parameters(), lr=1e-3)
for _ in range(100):
    optimizer.step()
```

ryanjulian commented:
Thanks Vitchyr!

I think you make excellent points and ask questions which are really worth addressing.

I initially found it confusing why you wouldn't see the value in this proposal, then I did my homework and realized that the problem is much smaller than we originally thought. I know you stopped enduring the pain of TensorFlow whereas I'm a PyTorch newbie.

tl;dr: This is mostly a TensorFlow problem.

## My new thoughts
* `pickle` is the universal interface
* Make an easy wrapper for `Serializable`-style pickling of non-pickleable classes from external libraries (this can be somewhat magical)
* Delete `Serializable` use in environment classes and just use `EzPickle` from `gym`
* Decouple `Parameterized` and `Serializable`
* Remove use of `Serializable` in TensorFlow in favor of a proper `Model` class which handles pickling (similar to `nn.Module` from PyTorch)

## A second look at environments -- not a huge problem
I just took a second look at `gym` and realized that all their environments are now pickleable, which is cool. They are using `EzPickle` which is OpenAI's copy of `Serializable` (it's possible that it was originally copied from rllab's `Serializable`--there was a lot of overlap for a while).

```python
>>> e = gym.make('Ant-v2')
>>> e.unwrapped.__reduce__()
(<function _reconstructor at 0x7fa1fb54ee18>, (<class 'gym.envs.mujoco.ant.AntEnv'>, <class 'object'>, None), {'_ezpickle_args': (), '_ezpickle_kwargs': {}})
```

I also checked `dm_control` and they implement pickling directly in their MuJoCo bindings.

Writing custom environments is fairly uncommon, so I think it would be reasonable just to tell people they need to make their environments pickleable (using `Serializable`, `EzPickle`, or whatever else).

Since we have standardized garage on the `gym.Env` interface, I would probably advocate that we move to `EzPickle` for custom environment classes.

For non-pickleable environments (e.g. I don't know if ALE supports pickling), it's simple enough for the `runner` framework from my example to provide a `Serializable`-like wrapper or decorator.

### Preliminaries
Assume I have some policy architecture which can be thought of as a tuple `(network_graph, hyperparameters)`. Currently we concretely represent that as a class (e.g. `GaussianMLPPolicy`) and constructor parameters, so it looks like `(GaussianMLPPolicy, (args, kwargs))`.

I am advocating that we force that all to be wrapped in an `Agent` class eventually since new agents are getting pretty complicated these days, and I will continue my example under this assumption.

```python
# Policy class
class GaussianMLPPolicy(Serializable, ...):

    def __init__(self, *args, **kwargs):
        Serializable.quick_init(self, locals())
        ...


# Agent class
class PPOAgent(Serializable, ...):

    def __init__(self, policy, baseline, ...):
        Serializable.quick_init(self, locals())
        ...




# Launcher file
policy = GaussianMLPPolicy(hidden_sizes=(100, 200, 300), hidden_nonlinearity=tf.nn.tanh)
baseline = GaussianMLPBaseline(hidden_sizes=(10, 20, 30))
agent = PPOAgent(policy=policy, baseline=baseline, ...)
environment = gym.make('Cartpole-v1')

run_experiment(agent, environment, iterations=1e3)
```

## Why I (still) want (something like) `Serializable`
I want to transmit configured agents across process boundaries, networks, or disks in a lightweight and standardized way.

I agree that `pickle` is _the_ interface for this. The question is what is pickleable and how.

__Multiprocess and distributed sampling__
For instance, in distrubuted or multiprocess sampling, I want to start `n` sampler workers, either on a single machine as separate processes, or across a network. I need a way to serialize the (neural network-backed) agent into bytes, and deserialize those bytes into an agent on the other side. This process could include allocating a lot of GPU memory, for instance.

__Checkpointing to disk__
This is similar to above, but the target is a disk, and I might unpickle my agent at any time in the future.

I realize that PyTorch provides tools for the multiprocessing and distributed cases, and the PyTorch sampler would probably use those. TensorFlow provides no such tools so we have to roll out own--maybe in TF 2.0??

### Pickling in TensorFlow
If you want to pickle a TensorFlow model, you will have to roll your own.

```python
>>> import tensorflow as tf
>>> sess = tf.InteractiveSession()
>>> input = tf.placeholder(tf.float32, shape=(1024,1024))
>>> from garage.tf.core.networks import mlp
>>> m = mlp(input, 32, (16, 32, 64), "my_mlp")
>>> m
<tf.Tensor 'my_mlp/output/BiasAdd:0' shape=(1024, 32) dtype=float32>
>>> import pickle
>>> mm = pickle.loads(pickle.dumps(m))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't pickle _thread.RLock objects
```

This is for a single MLP, but would still apply if `m` were a member of a class. `tf.keras.Model` does not provide pickling either.

```python
>>> inputs = tf.keras.Input(shape=(3,))
>>> x = tf.keras.layers.Dense(4, activation=tf.nn.relu)(inputs)
>>> outputs = tf.keras.layers.Dense(5, activation=tf.nn.softmax)(x)
>>> model = tf.keras.Model(inputs=inputs, outputs=outputs)
>>> model
<tensorflow.python.keras.engine.training.Model object at 0x7f7b83de5390>
>>> pickle.loads(pickle.dumps(model))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't pickle _thread.RLock objects
>>>
```

### Pickling in PyTorch
```python
In [1]: import torch
   ...: import torch.utils.data
   ...: from torch import nn, optim
   ...: from torch.nn import functional as F
   ...: from torchvision import datasets, transforms
   ...: from torchvision.utils import save_image
   ...: 
   ...: 

In [2]: class VAE(nn.Module):
   ...:     def __init__(self):
   ...:         super(VAE, self).__init__()
   ...: 
   ...:         self.fc1 = nn.Linear(784, 400)
   ...:         self.fc21 = nn.Linear(400, 20)
   ...:         self.fc22 = nn.Linear(400, 20)
   ...:         self.fc3 = nn.Linear(20, 400)
   ...:         self.fc4 = nn.Linear(400, 784)
   ...: 
   ...:     def encode(self, x):
   ...:         h1 = F.relu(self.fc1(x))
   ...:         return self.fc21(h1), self.fc22(h1)
   ...: 
   ...:     def reparameterize(self, mu, logvar):
   ...:         std = torch.exp(0.5*logvar)
   ...:         eps = torch.randn_like(std)
   ...:         return eps.mul(std).add_(mu)
   ...: 
   ...:     def decode(self, z):
   ...:         h3 = F.relu(self.fc3(z))
   ...:         return torch.sigmoid(self.fc4(h3))
   ...: 
   ...:     def forward(self, x):
   ...:         mu, logvar = self.encode(x.view(-1, 784))
   ...:         z = self.reparameterize(mu, logvar)
   ...:         return self.decode(z), mu, logvar
   ...:     

In [3]: model = VAE()

In [4]: import pickle

In [5]: pickle.loads(pickle.dumps(model))
Out[5]: 
VAE(
  (fc1): Linear(in_features=784, out_features=400, bias=True)
  (fc21): Linear(in_features=400, out_features=20, bias=True)
  (fc22): Linear(in_features=400, out_features=20, bias=True)
  (fc3): Linear(in_features=20, out_features=400, bias=True)
  (fc4): Linear(in_features=400, out_features=784, bias=True)
)

In [6]: mm = pickle.loads(pickle.dumps(model))

In [7]: mm
Out[7]: 
VAE(
  (fc1): Linear(in_features=784, out_features=400, bias=True)
  (fc21): Linear(in_features=400, out_features=20, bias=True)
  (fc22): Linear(in_features=400, out_features=20, bias=True)
  (fc3): Linear(in_features=20, out_features=400, bias=True)
  (fc4): Linear(in_features=400, out_features=784, bias=True)
)

In [9]: model
Out[9]: 
VAE(
  (fc1): Linear(in_features=784, out_features=400, bias=True)
  (fc21): Linear(in_features=400, out_features=20, bias=True)
  (fc22): Linear(in_features=400, out_features=20, bias=True)
  (fc3): Linear(in_features=20, out_features=400, bias=True)
  (fc4): Linear(in_features=400, out_features=784, bias=True)
)

In [11]: device = torch.device("cuda")

In [12]: model = VAE().to(device)

In [13]: mm = pickle.loads(pickle.dumps(model))

In [14]: mm
Out[14]: 
VAE(
  (fc1): Linear(in_features=784, out_features=400, bias=True)
  (fc21): Linear(in_features=400, out_features=20, bias=True)
  (fc22): Linear(in_features=400, out_features=20, bias=True)
  (fc3): Linear(in_features=20, out_features=400, bias=True)
  (fc4): Linear(in_features=400, out_features=784, bias=True)
)
```

krzentner commented:
On the topic of what `Serializable` **is**, I largely agree with @vitchyr. It shouldn't be an interface, it should just be a lightweight way of implementing the pickle interface. It's clear to me that putting `Serializable` anywhere near the root of the inheritance hierarchy is asking for trouble, but we've done that a few times.

I really do like just using the picklable interface whenever possible, since it's quite standard, well documented, and easy to use. If that's sufficient, then I'll withdraw this proposal.

Originally, I was concerned that sometimes we would have external types that aren't picklable, so it would be cleaner to pass around constructors to those types. However, that appears to not be much of a problem in practice. In the worst case, it's possible to construct a type which just calls a constructor when unpickled ([gist](https://gist.github.com/krzentner/abfcd00e3cc6e54d5ccf1a0b57eb5364)).

ryanjulian commented:
@krzentner I don't think there's much for Vitchyr and you to talk about because PyTorch doesn't need `Serializable`. Internal to garage we can just delete `Serializable` and go about our lives.

I think we should still come to a consensus about what to do about TF with @hartikainen since he maintains a popular TensorFlow/rllab-based repo (sac), which I hope he can some day rebase onto garage.

hartikainen commented:
> If you want to pickle a TensorFlow model, you will have to roll your own.

Looking at https://github.com/keras-team/keras/pull/11030, it seems like Keras actually nowadays supports pickling. However, this is not yet incorporated in `tf.keras`. Also, I'm not sure if that supports subclassing `tf.keras.Model` or not.

hartikainen commented:
Fyi: https://github.com/tensorflow/community/blob/024e81ce1f8795991043610cc2b734fb29f5a75c/rfcs/20181116-saved-model.md

ryanjulian commented:
Thanks Kristian!

krzentner commented:
This PR is no longer relevant, since we have the `Model` class in garage, and have found it fairly easy to use `__getstate__` and `__setstate__` to make anything else picklable, if necessary.
